src/Processors/Transforms/DeduplicationTokenTransforms.cpp

```diff
@@ -0,0 +1,205 @@
+#include <Processors/Transforms/DeduplicationTokenTransforms.h>
+
+#include <IO/WriteHelpers.h>
+
+#include <Common/logger_useful.h>
+#include <Common/Exception.h>
+#include <Common/SipHash.h>
+
+
+#include <fmt/core.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+void RestoreChunkInfosTransform::transform(Chunk & chunk)
+{
+    chunk.getChunkInfos().append(chunk_infos.clone());
+}
+
+namespace DeduplicationToken
+{
+
+String DB::DeduplicationToken::TokenInfo::getToken(bool enable_assert) const
+{
+    chassert(stage == VIEW_ID || !enable_assert);
+
+    String result;
+    result.reserve(getTotalSize());
+
+    for (const auto & part : parts)
+    {
+        if (!result.empty())
+            result.append(":");
+        result.append(part);
+    }
+
+    return result;
+}
+
+void DB::DeduplicationToken::TokenInfo::addPieceToInitialToken(String part)
+{
+    chassert(stage == INITIAL);
+    addTokenPart(std::move(part));
+}
+
+void DB::DeduplicationToken::TokenInfo::closeInitialToken()
+{
+    chassert(stage == INITIAL);
+    stage = VIEW_ID;
+}
+
+void TokenInfo::setUserToken(const String & token)
+{
+    chassert(stage == INITIAL);
+    addTokenPart(fmt::format("user-token-{}", token));
+    stage = SOURCE_BLOCK_NUMBER;
+}
+
+void TokenInfo::setSourceBlockNumber(size_t sbn)
+{
+    chassert(stage == SOURCE_BLOCK_NUMBER);
+    addTokenPart(fmt::format("source-number-{}", sbn));
+    stage = VIEW_ID;
+}
+
+void TokenInfo::setViewID(const String & id)
+{
+    chassert(stage == VIEW_ID);
+    addTokenPart(fmt::format("view-id-{}", id));
+    stage = VIEW_BLOCK_NUMBER;
+}
+
+void TokenInfo::setViewBlockNumber(size_t mvbn)
+{
+    chassert(stage == VIEW_BLOCK_NUMBER);
+    addTokenPart(fmt::format("view-block-{}", mvbn));
+    stage = VIEW_ID;
+}
+
+void TokenInfo::reset()
+{
+    stage = INITIAL;
+    parts.clear();
+}
+
+void TokenInfo::addTokenPart(String part)
+{
+    if (!part.empty())
+        parts.push_back(std::move(part));
+}
+
+size_t TokenInfo::getTotalSize() const
+{
+    if (parts.empty())
+        return 0;
+
+    size_t size = 0;
+    for (const auto & part : parts)
+        size += part.size();
+
+    return size + parts.size() - 1;
+}
+
+void CheckTokenTransform::transform(Chunk & chunk)
+{
+    auto token_info = chunk.getChunkInfos().get<TokenInfo>();
+
+    if (!token_info)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Chunk has to have DedupTokenInfo as ChunkInfo, {}", debug);
+
+    if (!must_be_present)
```

Comment:
Should we set token_info at all in this case?

Reply:
Token info is set always for inserts. After that it might be reset if needed.

For some tables it is not required because it decide about deduplication by themself. 

When `disable_deduplication_for_children=true` we turn off deduplication for depended tables by 
```
        pipeline.addTransform(std::make_shared<DeduplicationToken::ResetTokenTransform>(pipeline.getHeader()));
```
However for the main table it is still present and reset only for the depented tables.

But Im going to deprecate `disable_deduplication_for_children`. So I preferred to desable it when neded, instead create when needed. Because it almost always is needed.